src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+                d = self.factory.portal.login(
+                    self.blob_manager.credential,
+                    SMBMind(req.prev_session_id,
+                            self.blob_manager.credential.domain, self.addr),
+                    ISMBServer)
+                d.addCallback(self._cb_login, resp)
+                d.addErrback(self._eb_login, resp)
+            else:
+                resp.buffer = self.blob_manager.generateChallengeBlob()
+                self.session_setup_response(resp, STATUS_MORE_PROCESSING)
+
+    def _cb_login(self, t, resp):
+        _, self.avatar, self.logout_thunk = t
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(True)
+        log.debug("successful login")
+        self.session_setup_response(resp, STATUS_SUCCESS)
+
+    def _eb_login(self, failure, resp):
+        log.debug(failure.getTraceback())
+        resp.buffer = self.blob_manager.generateAuthResponseBlob(False)
+        self.session_setup_response(resp, STATUS_LOGON_FAILURE)
+
+    def session_setup_response(self, resp, ntstatus):
+        log.debug("session_setup_response")
+        resp.flags = 0
+        if self.blob_manager.credential == ANONYMOUS:
+            resp.flags |= SESSION_FLAG_IS_NULL
+        resp.size = 9
+        resp.offset = 72
+        resp.buflen = len(resp.buffer)
+        self.send_with_header(resp, 'session_setup', ntstatus)
+
+    def smb_logoff(self, req, resp):
+        assert req.size == 4
+        resp.size = 4
+        self.send_with_header(resp)
+        if self.logout_thunk:
+            d = maybeDeferred(self.logout_thunk)
+            d.addErrback(self._eb_logoff)
+
+    def _eb_logoff(self, f):
+        log.error(f.getTraceback())
+
+    def smb_tree_connect(self, req, resp):
+        if self.avatar is None:
+            self.error_response(STATUS_ACCESS_DENIED)
+            return
+        assert req.size == 9
+        path = req.buffer[req.offset - len(req) - 64:req.offset - len(req) -
+                          64 + req.buflen]
+        path = path.decode("utf-16le").split("\\")[-1]
+        try:
+            share = self.avatar.getShare(path)
+        except KeyError:
+            self.error_response(STATUS_BAD_NETWORK_NAME)
+            return
+        if IFilesystem.providedBy(share):
+            resp.share_type = SHARE_DISK
+            # FUTURE: select these values from share object
+            resp.flags = SHAREFLAG_MANUAL_CACHING
+            resp.capabilities = 0
+            resp.max_perms = (FILE_READ_DATA | FILE_WRITE_DATA
+                              | FILE_APPEND_DATA | FILE_READ_EA | FILE_WRITE_EA
+                              | FILE_DELETE_CHILD | FILE_EXECUTE
+                              | FILE_READ_ATTRIBUTES | FILE_WRITE_ATTRIBUTES
+                              | DELETE | READ_CONTROL | WRITE_DAC | WRITE_OWNER
+                              | SYNCHRONIZE)
+        elif IPipe.providedBy(share):
```

Comment:
Can it only be one? What happens if it provides more than one? Do we have any error reporting to the user?

Reply:
it makes no sense for a share to be more than one type

Reply:
The code now throws `AssertionError` if a share tries to implement more than one type.